### ISSUE TITLE
How to use logits_processors

### ISSUE THREAD

**Issue by shuaiwang2022** [vllm-project/vllm#1728]
Hello friends, how can I use the parameter logits_processors to modify the current logit?

**Comment by tolecy** [comment]
If you mean to modify the logits of some specific tokens, you can define a processing function about logits and pass this function into SamplingParams.

For example：
```
def process_token(token_ids, logits):
    logits[6623] = -9999.999 
    logits[4853] = -9999.999
    logits[630] = -9999.999
    logits[985] = -9999.999
    return logits

sampling_params = SamplingParams(
            n=1,
            temperature=temperature,
            top_p=top_p,
            use_beam_search=use_beam_search,
            stop=list(stop),
            max_tokens=max_new_tokens,
            top_k=top_k,
            presence_penalty=presence_penalty,
            frequency_penalty=frequency_penalty,
            best_of=best_of,
            logits_processors=[process_token]
        )
```

You can refer to it
https://github.com/vllm-project/vllm/blob/e19a64c7eff2085790dbf71851208fa2dd31ca4d/tests/samplers/test_sampler.py#L188

**Comment by hummingbird2030** [comment]
> If you mean to modify the logits of some specific tokens, you can define a processing function about logits and pass this function into SamplingParams.
> 
> For example：
> 
> ```
> def process_token(token_ids, logits):
>     logits[6623] = -9999.999 
>     logits[4853] = -9999.999
>     logits[630] = -9999.999
>     logits[985] = -9999.999
>     return logits
> 
> sampling_params = SamplingParams(
>             n=1,
>             temperature=temperature,
>             top_p=top_p,
>             use_beam_search=use_beam_search,
>             stop=list(stop),
>             max_tokens=max_new_tokens,
>             top_k=top_k,
>             presence_penalty=presence_penalty,
>             frequency_penalty=frequency_penalty,
>             best_of=best_of,
>             logits_processors=[process_token]
>         )
> ```
> 
> You can refer to it
> 
> https://github.com/vllm-project/vllm/blob/e19a64c7eff2085790dbf71851208fa2dd31ca4d/tests/samplers/test_sampler.py#L188

 I have a question about the usage of [process_token]. How does this function get the token_ids and logits?

**Comment by hmellor** [comment]
The list of `logits_processors` is looped over in here:
https://github.com/vllm-project/vllm/blob/c3af44722cff56bba5fc912c8e16d9de02dfb532/vllm/model_executor/layers/logits_processor.py#L88-L115

**Comment by ctscsu** [comment]
> If you mean to modify the logits of some specific tokens, you can define a processing function about logits and pass this function into SamplingParams.
> 
> For example：
> 
> ```
> def process_token(token_ids, logits):
>     logits[6623] = -9999.999 
>     logits[4853] = -9999.999
>     logits[630] = -9999.999
>     logits[985] = -9999.999
>     return logits
> 
> sampling_params = SamplingParams(
>             n=1,
>             temperature=temperature,
>             top_p=top_p,
>             use_beam_search=use_beam_search,
>             stop=list(stop),
>             max_tokens=max_new_tokens,
>             top_k=top_k,
>             presence_penalty=presence_penalty,
>             frequency_penalty=frequency_penalty,
>             best_of=best_of,
>             logits_processors=[process_token]
>         )
> ```
> 
> You can refer to it
> 
> https://github.com/vllm-project/vllm/blob/e19a64c7eff2085790dbf71851208fa2dd31ca4d/tests/samplers/test_sampler.py#L188


if i have a batch data. say batchsize 16.
but each data i want to design a different logits_processors.
how to do it?.
one idea is to set batchsize 1 .
but i need online infer .batch size must be larger

**Comment by Dan-wanna-M** [comment]
> > If you mean to modify the logits of some specific tokens, you can define a processing function about logits and pass this function into SamplingParams.
> > For example：
> > ```
> > def process_token(token_ids, logits):
> >     logits[6623] = -9999.999 
> >     logits[4853] = -9999.999
> >     logits[630] = -9999.999
> >     logits[985] = -9999.999
> >     return logits
> > 
> > sampling_params = SamplingParams(
> >             n=1,
> >             temperature=temperature,
> >             top_p=top_p,
> >             use_beam_search=use_beam_search,
> >             stop=list(stop),
> >             max_tokens=max_new_tokens,
> >             top_k=top_k,
> >             presence_penalty=presence_penalty,
> >             frequency_penalty=frequency_penalty,
> >             best_of=best_of,
> >             logits_processors=[process_token]
> >         )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can refer to it
> > https://github.com/vllm-project/vllm/blob/e19a64c7eff2085790dbf71851208fa2dd31ca4d/tests/samplers/test_sampler.py#L188
> 
> if i have a batch data. say batchsize 16. but each data i want to design a different logits_processors. how to do it?. one idea is to set batchsize 1 . but i need online infer .batch size must be larger

I do not think it is possible in current version of vllm. The source code indicates the same `logits_processor` list is applied on all sequences in one batch

**Comment by nivibilla** [comment]
Hey, if I want to simply extract all the logits (size vocab size) how can I do this?

**Comment by const-q** [comment]
> Hey, if I want to simply extract all the logits (size vocab size) how can I do this?

If this is your outputs generator:

`        outputs = model.generate(
            input_ids=input_ids,
            attention_mask=attention_mask,
            max_new_tokens=128,
            do_sample=True,
            top_k=False,
            eos_token_id=tokenizer.eos_token_id,
            no_repeat_ngram_size=2,
            output_scores=True,
            return_dict_in_generate=True
        )`

Just make sure:

           `output_scores=True,
            return_dict_in_generate=True`

Then, all the logits will be stored in 'outputs.scores'

So, you can be free to manipulate the data however you want.

After generator a sequence, however, you will only get the last tensor of logits, which may look like this:

`(tensor([[59.2786, 60.0318, 57.9207,  ...,43.4708, 48.8161, 38.0972]]))`

Depending on if you are doing batch processing, in which case, you'll have additional brackets to delineate the batches like me.

If you want to access the logits of the whole sequence (so per token), then it would be best to use a for loop.

I hope this helps and I can answer if you have further questions since I'm working on a logit project right now.

**Comment by renhouxing** [comment]
Hi, I want to confirm whether vllm service can define a custom logit processor when requesting or starting the service.

**Comment by dinhngoc267** [comment]
> token_ids



> > Hey, if I want to simply extract all the logits (size vocab size) how can I do this?
> 
> If this is your outputs generator:
> 
> ```
>        `outputs = model.generate(
>         input_ids=input_ids,
>         attention_mask=attention_mask,
>         max_new_tokens=128,
>         do_sample=True,
>         top_k=False,
>         eos_token_id=tokenizer.eos_token_id,
>         no_repeat_ngram_size=2,
>         output_scores=True,
>         return_dict_in_generate=True
>     )`
> ```
> 
> Just make sure:
> 
> ```
>        `output_scores=True,
>         return_dict_in_generate=True`
> ```
> 
> Then, all the logits will be stored in `outputs.scores`
> 
> So, you can be free to manipulate the data however you want.
> 
> After generating a sequence, however, you will only get the last tensor of logits, which may look like this:
> 
> `(tensor([[59.2786, 60.0318, 57.9207, ...,43.4708, 48.8161, 38.0972]]))`
> 
> Depending on if you are doing batch processing, in which case, you'll have additional brackets to delineate the batches like me.
> 
> Since you want to access logits of the whole `vocab_size` (so per token), then it would be best to use a for loop.
> 
> I hope this helps and I can answer if you have further questions since I'm working on a logits project right now.

Hi, why the token_ids, not the same between each output? 

```
tensor([[ 7.7500,  7.8571,  3.0357,  ..., -3.1562, -3.1562, -3.1562]],
       device='cuda:0')
tensor([[151644,   8948,    198,  ...,    582,    614,   2525]],  # token_ids tensor of time step i-th
       device='cuda:0')
torch.Size([1, 152064])
tensor([[ 4.6562,  6.1310, -0.5147,  ..., -3.0312, -3.0312, -3.0312]],
       device='cuda:0')
tensor([[151644,   8948,    198,  ...,    614,   2525,    311]],  # token_ids tensor of time step i+1 th
       device='cuda:0')
torch.Size([1, 152064])
```

Then I cannot use the same mask apply for all the time step. loop each token_ids cost too much time.

**Comment by shuaiwang2022** [comment]
def process_token(token_ids, logits):
    logits[6623] = -9999.999 
    logits[4853] = -9999.999
    logits[630] = -9999.999
    logits[985] = -9999.999
    return logits

sampling_params = SamplingParams(
            n=1,
            temperature=temperature,
            top_p=top_p,
            use_beam_search=use_beam_search,
            stop=list(stop),
            max_tokens=max_new_tokens,
            top_k=top_k,
            presence_penalty=presence_penalty,
            frequency_penalty=frequency_penalty,
            best_of=best_of,
            logits_processors=[process_token]
        )

How to use process_token? How can I get the logit of each token?

**Comment by dinhngoc267** [comment]
> process_token

this process_token recieve the logits of each step. has shape (vocabulary, 1) each element in this array is your "logit of each token"

**Comment by shuaiwang2022** [comment]
Thank you so much

**Comment by lionsheep0724** [comment]
Does this implementation decrease throughputs?I had some degradation of TPS when using this. Any similar issues here?

**Comment by Kenwwww** [comment]
> Hi, I want to confirm whether vllm service can define a custom logit processor when requesting or starting the service.

do you have any results？

**Comment by dinhngoc267** [comment]
> > Hi, I want to confirm whether vllm service can define a custom logit processor when requesting or starting the service.
> 
> do you have any results？

Can u check this link: 

https://github.com/dinhngoc267/ban-chinese-in-Qwen/blob/main/vllm.py

**Comment by Kenwwww** [comment]
> > > Hi, I want to confirm whether vllm service can define a custom logit processor when requesting or starting the service.
> > 
> > 
> > do you have any results？
> 
> Can u check this link:
> 
> https://github.com/dinhngoc267/ban-chinese-in-Qwen/blob/main/vllm.py

yeah，I have defined the processor,  while I don‘t know how to initial it by  vllm   openai server   with "vllm serve" , and how to pass the content in client

**Comment by quangvinh2110** [comment]
> > > > Hi, I want to confirm whether vllm service can define a custom logit processor when requesting or starting the service.
> > > 
> > > 
> > > do you have any results？
> > 
> > 
> > Can u check this link:
> > https://github.com/dinhngoc267/ban-chinese-in-Qwen/blob/main/vllm.py
> 
> yeah，I have defined the processor, while I don‘t know how to initial it by vllm openai server with "vllm serve" , and how to pass the content in client

any solution to pass logit processor to vllm api?
